helix-core/src/main/java/org/apache/helix/controller/stages/TaskGarbageCollectionStage.java

```diff
@@ -93,12 +97,18 @@ public void process(ClusterEvent event) throws Exception {
           .equals(TaskUtil.WORKFLOW_CONTEXT_KW)) {
         // Find workflows that need to be purged
         workflowsToBePurged.add(entry.getKey());
+      } else if (jobConfig == null && entry.getValue() != null && entry.getValue().getId()
```

Comment:
Could this job still exist in WF context? Do we need to scan WF context and clean them up?

Reply:
This follows what we do for expired jobs. Note that in expired jobs we do not do that as well because main pipeline is responsible to update workflow context. The reason has been also explained in the comments:
```
      // Update workflow context will be in main pipeline not here. Otherwise, it will cause
      // concurrent write issue. It is possible that jobs got purged but there is no event to
      // trigger the pipeline to clean context.
```